### PR TITLE
[MCP] Optimize copies when src is used during backward propagation

### PR DESCRIPTION
**Pull request by vladimirradosavljevic**
Before this patch, redundant COPY couldn't be removed for the following case:
```
  $R0 = OP ...
  ... // Read of %R0
  $R1 = COPY killed $R0
```
This patch adds support for tracking the users of the source register during backward propagation, so that we can remove the redundant COPY in the above case and optimize it to:
```
  $R1 = OP ...
  ... // Replace all uses of %R0 with $R1
```

### CODE REVIEW

File: llvm/test/CodeGen/Mips/micromips-gp-rc.ll [hunk 1]
```diff
@@ -1,18 +0,0 @@
...
-entry:
-  %0 = load i32, ptr @g, align 4
-  tail call void @exit(i32 signext %0)
-  unreachable
-}
-
-; Function Attrs: noreturn
-declare void @exit(i32 signext)
-
-; CHECK: move $gp, ${{[0-9]+}}
```
Comment (vladimirradosavljevic) [inline llvm/test/CodeGen/Mips/micromips-gp-rc.ll:17]: Removed this test, since this copy instruction is optimized.

**Comment by vladimirradosavljevic** [comment]
@qcolombet @arsenm @topperc PTAL.

File: llvm/test/CodeGen/Mips/micromips-gp-rc.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/Mips/micromips-gp-rc.ll:17]: Don't remove tests

Comment (vladimirradosavljevic) [reply llvm/test/CodeGen/Mips/micromips-gp-rc.ll:17]: I restored this test, and updated checks.

File: llvm/lib/CodeGen/MachineCopyPropagation.cpp [hunk 2]
```diff
@@ -667,6 +707,26 @@ bool MachineCopyPropagation::hasOverlappingMultipleDef(
   return false;
 }
 
+/// Return true if it is safe to update the users of the source register of the
+/// copy.
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineCopyPropagation.cpp:711]: Document parameters?

**Review by arsenm — COMMENTED** [review]
Is one of these test changes a dedicated MIR test for this situation?

**Comment by qcolombet** [comment]
> Is one of these test changes a dedicated MIR test for this situation?

Having a quick glance at the patch I don't think it is.

Could you add a few tests that checks specifically for the new pattern?
E.g.,
```
r0 = ..
// one of the use of r0 is not renameable
r1 = copy r0
```
and also
```
r0 = ..
// one of the use of r0 does not accept r1
r1 = copy r0
```
and so on.

Essentially add explicit test coverage with focused mir tests.

**Comment by qcolombet** [comment]
Note: I think you support all these cases, but I'd like to see them written down.

File: llvm/lib/CodeGen/MachineCopyPropagation.cpp (hunk 2 above)
Comment (vladimirradosavljevic) [reply llvm/lib/CodeGen/MachineCopyPropagation.cpp]: Done.

**Comment by vladimirradosavljevic** [comment]
@arsenm @qcolombet I added dedicated MIR test to cover 3 scenarios:
1. Normal case where transformation is performed.
2. Where use is not renamable.
3. Implicit use.

**Comment by vladimirradosavljevic** [comment]
Ping.

**Review by qcolombet — APPROVED** [review]
LGTM

**Comment by vladimirradosavljevic** [comment]
Rebased and updated checks in `midpoint-int.ll`.
@qcolombet Thanks for the review! Could you please merge this PR on my behalf?

**Pull request merged**